### PR TITLE
Allow image registry override for all keda components

### PR DESCRIPTION
**Pull request by QuentinBisson**
This PR allow override of the image registry of all keda components so we can use custom image or pull through caching solutions

### Checklist

- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/charts/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] README is updated with new configuration values *(if applicable)* [learn more](https://github.com/kedacore/charts/blob/main/CONTRIBUTING.md#documentation)
- [ ] 
Fixes #

### CODE REVIEW

File: keda/Chart.yaml [hunk 1]
```diff
@@ -8,11 +8,11 @@ kubeVersion: ">=v1.23.0-0"
 
 # This is the chart version. This version number should be incremented each time you make changes
 # to the chart and its templates, including the app version.
-version: 2.12.0
+version: 2.12.1
```
Comment (tomkerkhove) [inline keda/Chart.yaml:11]: Can you please revert these changes and remove the new version?

File: keda/values.yaml [hunk 2]
```diff
@@ -2,20 +2,31 @@
...
+    # -- Global image registry of KEDA components
+    registry: null
+
 image:
   keda:
+    # -- Image registry of KEDA operator
+    registry: ghcr.io
     # -- Image name of KEDA operator
-    repository: ghcr.io/kedacore/keda
+    repository: kedacore/keda
```
Comment (tomkerkhove) [inline keda/values.yaml:15]: These are breaking changes, we'll need to revert it and find another way to not break people

**Review by tomkerkhove — CHANGES_REQUESTED** [review]

File: keda/Chart.yaml (hunk 1 above)
Comment (QuentinBisson) [reply keda/Chart.yaml:11]: Could you explain why? That's what is explained in Contributing.md

File: keda/values.yaml (hunk 2 above)
Comment (QuentinBisson) [reply keda/values.yaml:15]: I'm afraid that doing in another way would diverge il much with what all the other charts in the wild do though. I'm fine if it makes out in 2.13

File: keda/Chart.yaml (hunk 1 above)
Comment (tomkerkhove) [reply keda/Chart.yaml:11]: That is just overall docs, not every PR is supposed to do that. I'll update the docs next week

**Comment by QuentinBisson** [comment]
I have an idea, i'll change that pr monday

**Comment by QuentinBisson** [comment]
Local testing:

- default values renders:

```sh
helm template keda | grep image: | grep keda
          image: "ghcr.io/kedacore/keda:2.12.0"
          image: "ghcr.io/kedacore/keda-metrics-apiserver:2.12.0"
          image: "ghcr.io/kedacore/keda-admission-webhooks:2.12.0"
```

- edited component registries (require changing the repository but that ensures no breaking changes)

```yaml
global:
  image:
    # -- Global image registry of KEDA components
    registry: null

image:
  keda:
    # -- Image registry of KEDA operator
    registry: docker.io
    # -- Image name of KEDA operator
    repository: kedacore/keda
    # -- Image tag of KEDA operator. Optional, given app version of Helm chart is used by default
    tag: ""
  metricsApiServer:
    # -- Image registry of KEDA Metrics API Server
    registry: docker.io
    # -- Image name of KEDA Metrics API Server
    repository: kedacore/keda-metrics-apiserver
    # -- Image tag of KEDA Metrics API Server. Optional, given app version of Helm chart is used by default
    tag: ""
  webhooks:
    # -- Image registry of KEDA admission-webhooks
    registry: docker.io
    # -- Image name of KEDA admission-webhooks
    repository: kedacore/keda-admission-webhooks
    # -- Image tag of KEDA admission-webhooks . Optional, given app version of Helm chart is used by default
    tag: ""
  # -- Image pullPolicy for all KEDA components
  pullPolicy: Always
```

Renders:

```sh
helm template keda | grep image: | grep keda
          image: "docker.io/kedacore/keda:2.12.0"
          image: "docker.io/kedacore/keda-metrics-apiserver:2.12.0"
          image: "docker.io/kedacore/keda-admission-webhooks:2.12.0"
```

- using global with repository change:
```yaml
global:
  image:
    # -- Global image registry of KEDA components
    registry: quay.io

image:
  keda:
    # -- Image registry of KEDA operator
    registry: docker.io
    # -- Image name of KEDA operator
    repository: kedacore/keda
    # -- Image tag of KEDA operator. Optional, given app version of Helm chart is used by default
    tag: ""
  metricsApiServer:
    # -- Image registry of KEDA Metrics API Server
    registry: docker.io
    # -- Image name of KEDA Metrics API Server
    repository: kedacore/keda-metrics-apiserver
    # -- Image tag of KEDA Metrics API Server. Optional, given app version of Helm chart is used by default
    tag: ""
  webhooks:
    # -- Image registry of KEDA admission-webhooks
    registry: docker.io
    # -- Image name of KEDA admission-webhooks
    repository: kedacore/keda-admission-webhooks
    # -- Image tag of KEDA admission-webhooks . Optional, given app version of Helm chart is used by default
    tag: ""
  # -- Image pullPolicy for all KEDA components
  pullPolicy: Always
```

Renders:

```sh
helm template keda | grep image: | grep keda
          image: "quay.io/kedacore/keda:2.12.0"
          image: "quay.io/kedacore/keda-metrics-apiserver:2.12.0"
          image: "quay.io/kedacore/keda-admission-webhooks:2.12.0"
```

**Comment by QuentinBisson** [comment]
@tomkerkhove what do you think about this solution?

**Comment by tomkerkhove** [comment]
Fine by me - Good idea.

However, we still need to remove the new artifact

**Comment by QuentinBisson** [comment]
@tomkerkhove I removed the artifact

**Review by tomkerkhove — APPROVED** [review]

**Comment by tomkerkhove** [comment]
Thanks! Sorry for the delay in merging!

**Comment by mehraneftekhari99** [comment]
The changes are great however they're not pushed to the helm chart registry.
The version in master branch (which includes overridable registry) is 2.12.1. However when you get the exact same version via helm cli, it doesn't have the change!
Something is wrong maybe in build process etc. just wanted to let you know since I spent some time figuring out why it's not working for me.

**Comment by JorTurFer** [comment]
Hi @mehraneftekhari99 
Thanks for reporting it! We haven't cut a release with that feature yet, we will ship it as part of v2.13 during next days
